{
  "code": 0,
  "subdomains": {
    "unexpected": {
      "code": 0,
      "errors": {
        "unknown_error": {
          "code": 1,
          "message": "Unknown error: %s.",
          "class": "InternalError"
        },
        "timeout": {
          "code": 2,
          "message": "[FATAL] Service \"%s[%s]\" failed to init within %sms",
          "class": "InternalError"
        }
      }
    },
    "external_services": {
      "code": 1,
      "errors": {
        "service_unavailable": {
          "code": 1,
          "message": "Service unavailable: %s.",
          "class": "ServiceUnavailableError"
        }
      }
    },
    "hotelclerk" : {
      "code": 2,
      "errors": {
        "room_id_not_exists": {
          "code": 1,
          "message": "The room Id \"%s\" does not exist.",
          "class": "NotFoundError"
        },
        "subscription_not_found": {
          "code": 2,
          "message": "No subscription found on index %s and collection %s.",
          "class": "NotFoundError"
        },
        "rooms_attribute_type": {
          "code": 3,
          "message": "The rooms attribute must be an array.",
          "class": "BadRequestError"
        },
        "incorrect_state_value": {
          "code": 4,
          "message": "Incorrect value for the 'state' parameter. Expected: all, done or pending. Got: %s.",
          "class": "BadRequestError"
        },
        "incorrect_scope_value": {
          "code": 5,
          "message": "Incorrect value for the 'scope' parameter. Expected: all, done or pending. Got: %s.",
          "class": "BadRequestError"
        },
        "incorrect_users_value": {
          "code": 6,
          "message": "Incorrect value for the 'users' parameter. Expected: all, done or pending. Got: %s.",
          "class": "BadRequestError"
        },
        "missing_subscription_parameter": {
          "code": 7,
          "message": "Cannot subscribe without an index and a collection.",
          "class": "BadRequestError"
        },
        "limit_minterms_reached": {
          "code": 8,
          "message": "Unable to subscribe: maximum number of minterms exceeded (max %s, received %s).",
          "class": "SizeLimitError"
        },
        "limit_unique_rooms_reached": {
          "code": 9,
          "message": "Unable to subscribe: maximum number of unique rooms reached.",
          "class": "SizeLimitError"
        },
        "cant_unsubscribe_unknown_user": {
          "code": 10,
          "message": "Unsubscribe error: no subscription found for that user.",
          "class": "NotFoundError"
        },
        "cant_unsubscribe_user_not_subscribed": {
          "code": 11,
          "message": "Unsubscribe error: not subscribed to %s.",
          "class": "NotFoundError"
        },
        "cant_unsubscribe_unknown_room": {
          "code": 12,
          "message": "Unsubscribe error: room %s not found.",
          "class": "InternalError"
        }
      }
    },
    "janitor": {
      "code": 3,
      "errors": {
        "import_data" : {
          "code": 1,
          "message" : "Some data was not imported for %s/%s (%s/%s).",
          "class": "PartialError"
        },
        "dump_already_generated": {
          "code": 2,
          "message": "A dump is already being generated. Skipping.",
          "class": "PreconditionError"
        },
        "reading_log_directory": {
          "code": 3,
          "message": "Cannot read log directory '%s' : %s.",
          "class": "InternalError"
        }
      }
    },
    "vault": {
      "code": 4,
      "errors": {
        "decrypt_secrets": {
          "code": 1,
          "message": "Cannot decrypt secrets: %s.",
          "class": "InternalError"
        },
        "vault_key_not_found": {
          "code": 2,
          "message": "Cannot find vault key. Aborting.",
          "class": "NotFoundError"
        },
        "secrets_must_be_object": {
          "code": 3,
          "message": "Secrets must be an object.",
          "class": "PreconditionError"
        }
      }
    },
    "plugins": {
      "code": 5,
      "errors": {
        "invalid_plugin_name": {
          "code": 1,
          "message": "[%s] Invalid plugin name. The name must be comprised only of lowercased letters, numbers, hyphens and underscores.",
          "class": "PluginImplementationError"
        },
        "invalid_privileged_property": {
          "code": 2,
          "message": "[%s] Invalid \"privileged\" property: expected a boolean, got a %s.",
          "class": "PluginImplementationError"
        },
        "missing_package_json": {
          "code": 3,
          "message": "[%s] No package.json file found.",
          "class": "PluginImplementationError"
        },
        "missing_name_property_in_package_json": {
          "code": 4,
          "message": "[%s] No \"name\" property provided in package.json.",
          "class": "PluginImplementationError"
        },
        "invalid_user_object": {
          "code": 5,
          "message": "You must provide a valid User object when adding context with as().",
          "class": "PluginImplementationError"
        },
        "collection_not_specified": {
          "code": 6,
          "message": "The collection must be specified.",
          "class": "PluginImplementationError"
        },
        "callback_argument_expected": {
          "code": 7,
          "message": "Invalid argument: Expected callback to be a function, received \"%s\".",
          "class": "PluginImplementationError"
        },
        "missing_request_object": {
          "code": 8,
          "message": "Invalid argument: a Request object must be supplied.",
          "class": "PluginImplementationError"
        },
        "invalid_custom_event_name": {
          "code": 9,
          "message": "Custom event invalid name (%s). Colons are not allowed in custom events.",
          "class": "PluginImplementationError"
        },
        "missing_request_data_or_object": {
          "code": 10,
          "message": "A Request object and/or request data must be provided.",
          "class": "PluginImplementationError"
        },
        "invalid_strategy_registration": {
          "code": 11,
          "message": "[%s] Strategy %s: dynamic strategy registration can only be done using an \"authenticator\" option (see https://tinyurl.com/y7boozbk).",
          "class": "PluginImplementationError"
        },
        "plugin_initialization_failed": {
          "code": 12,
          "message": "Something went wrong during initialization of \"%s\" plugin.",
          "class": "PluginImplementationError"
        },
        "strategy_description_type": {
          "code": 13,
          "message": "%s expected the strategy description to be an object, got: %s.",
          "class": "PluginImplementationError"
        },
        "methods_property_type": {
          "code": 14,
          "message": "%s expected a \"methods\" property of type \"object\", got: %s.",
          "class": "PluginImplementationError"
        },
        "methodname_property_type": {
          "code": 15,
          "message": "%s expected a \"%s\" property of type \"string\", got: %s.",
          "class": "PluginImplementationError"
        },
        "invalid_strategy_method": {
          "code": 16,
          "message": "%s the strategy method \"%s\" must point to an exposed function.",
          "class": "PluginImplementationError"
        },
        "invalid_property_type": {
          "code": 17,
          "message": "%s expected the \"%s\" property to be of type \"string\", got: %s.",
          "class": "PluginImplementationError"
        },
        "missing_config_property": {
          "code": 18,
          "message": "%s expected a \"config\" property of type \"object\", got: %s.",
          "class": "PluginImplementationError"
        },
        "cannot_set_ctor_and_authenticator": {
          "code": 19,
          "message": "%s the \"authenticator\" and \"constructor\" parameters cannot both be set.",
          "class": "PluginImplementationError"
        },
        "invalid_constructor_property_value": {
          "code": 20,
          "message": "%s invalid \"constructor\" property value: constructor expected.",
          "class": "PluginImplementationError"
        },
        "authenticator_property_type": {
          "code": 21,
          "message": "%s expected an \"authenticator\" property of type \"string\", got: %s.",
          "class": "PluginImplementationError"
        },
        "unknown_authenticator_value": {
          "code": 22,
          "message": "%s unknown authenticator value: %s.",
          "class": "PluginImplementationError"
        },
        "expected_object_type": {
          "code": 23,
          "message": "%s expected the \"%s\" property to be of type \"object\", got: %s.",
          "class": "PluginImplementationError"
        },
        "invalid_fields_property_type": {
          "code": 24,
          "message": "%s expected the \"fields\" property to be of type \"array\", got: %s.",
          "class": "PluginImplementationError"
        },
        "plugin_threw_non_kuzzle_error": {
          "code": 25,
          "message": "Plugin %s pipe for event '%s' threw a non-Kuzzle error: %s.",
          "class": "PluginImplementationError"
        },
        "verify_dont_return_promise": {
          "code": 26,
          "message": "%s expected the \"verify\" to return a Promise, got: %s.",
          "class": "PluginImplementationError"
        },
        "invalid_authentication_strategy_result": {
          "code": 27,
          "message": "%s invalid authentication strategy result.",
          "class": "PluginImplementationError"
        },
        "invalid_authentication_kuid": {
          "code": 28,
          "message": "%s invalid authentication kuid returned: expected a string, got a %s.",
          "class": "PluginImplementationError"
        },
        "unkown_kuzzle_user_identifier": {
          "code": 29,
          "message": "%s returned an unknown Kuzzle user identifier.",
          "class": "PluginImplementationError"
        },
        "cannot_remove_others_plugin_strategy": {
          "code": 30,
          "message": "Cannot remove strategy %s: owned by another plugin.",
          "class": "PluginImplementationError"
        },
        "cannot_remove_unexistant_strategy": {
          "code": 31,
          "message": "Cannot remove strategy %s: strategy does not exist.",
          "class": "PluginImplementationError"
        },
        "incorrect_controller_description_type": {
          "code": 32,
          "message": "%s Incorrect controller description type (expected object, got: \"%s\").",
          "class": "PluginImplementationError"
        },
        "unknown_property_key_in_route_definition": {
          "code": 33,
          "message": "%s Unknown property \"%s\" in route definition. %s",
          "class": "PluginImplementationError"
        },
        "key_cannot_be_empty_string": {
          "code": 34,
          "message": "%s \"%s\" must be a non-empty string.",
          "class": "PluginImplementationError"
        },
        "undefined_controller": {
          "code": 35,
          "message": "Undefined controller \"%s\". %s",
          "class": "PluginImplementationError"
        },
        "undefined_action": {
          "code": 36,
          "message": "Undefined action \"%s\". %s",
          "class": "PluginImplementationError"
        },
        "http_verb_not_allowed": {
          "code": 37,
          "message": "%s Only following http verbs are allowed: \"%s\". %s",
          "class": "PluginImplementationError"
        },
        "strategies_plugin_property_empty": {
          "code": 38,
          "message": "%s the exposed \"strategies\" plugin property must be a non-empty object.",
          "class": "PluginImplementationError"
        },
        "authenticators_plugin_property_not_an_object": {
          "code": 39,
          "message": "%s the exposed \"authenticators\" plugin property must be of type \"object\".",
          "class": "PluginImplementationError"
        },
        "invalid_authenticator": {
          "code": 40,
          "message": "%s invalid authenticator %s: expected a constructor.",
          "class": "PluginImplementationError"
        },
        "unable_to_load_plugin_from_directory": {
          "code": 41,
          "message": "Unable to load plugins from directory \"%s\"; %s.",
          "class": "PluginImplementationError"
        },
        "unable_to_load_plugin_from_path": {
          "code": 42,
          "message": "Unable to load plugin from path \"%s\"; %s.",
          "class": "PluginImplementationError"
        },
        "plugin_is_not_a_constructor": {
          "code": 43,
          "message": "Plugin %s is not a constructor.",
          "class": "PluginImplementationError"
        },
        "init_method_not_found": {
          "code": 44,
          "message": "[%s] No \"init\" method found.",
          "class": "PluginImplementationError"
        },
        "privileged_mode_not_supported": {
          "code": 45,
          "message": "The plugin \"%s\" is configured to run in privileged mode, but it does not seem to support it.",
          "class": "PluginImplementationError"
        },
        "privileged_mode_not_setted": {
          "code": 46,
          "message": "The plugin \"%s\" needs to run in privileged mode to work, you have to explicitly set \"privileged: true\" in its configuration.",
          "class": "PluginImplementationError"
        },
        "plugin_name_already_exists": {
          "code": 47,
          "message": "A plugin named %s already exists",
          "class": "PluginImplementationError"
        },
        "invalid_user_object_provided": {
          "code": 48,
          "message": "FunnelProtocol.constructor: Invalid User object \"%s\"",
          "class": "InternalError"
        }
      }
    },
    "validation": {
      "code": 6,
      "errors": {
        "missing_typename_property": {
          "code": 1,
          "message": "The typeName property must be defined in the validation type object.",
          "class": "PluginImplementationError"
        },
        "missing_function_validate": {
          "code": 2,
          "message": "The type %s must implement the function 'validate'.",
          "class": "PluginImplementationError"
        },
        "missing_function_validatefieldspecification": {
          "code": 3,
          "message": "The type %s must implement the function 'validateFieldSpecification'.",
          "class": "PluginImplementationError"
        },
        "missing_function_getstrictness": {
          "code": 4,
          "message": "The allowing children type %s must implement the function 'getStrictness'.",
          "class": "PluginImplementationError"
        },
        "type_already_defined": {
          "code": 5,
          "message": "The type %s is already defined.",
          "class": "PluginImplementationError"
        },
        "object_format_error": {
          "code": 6,
          "message": "All levels of an object have to be defined in the specification.",
          "class": "PreconditionError"
        },
        "type_not_allowed": {
          "code": 7,
          "message": "The field type %s is not allowed to have children fields.",
          "class": "PreconditionError"
        },
        "parent_field_not_defined": {
          "code": 8,
          "message": "The parent field of the field \"%s\" is not defined.",
          "class": "PreconditionError"
        },
        "errorcontext_document": {
          "code": 9,
          "message": "Document: %s.",
          "class": "BadRequestError"
        },
        "invalid_formats_option": {
          "code": 10,
          "message": "Invalid \"formats\" option: must be a non-empty array.",
          "class": "PreconditionError"
        },
        "unrecognized_format_name": {
          "code": 11,
          "message": "Unrecognized format name%s: %s.",
          "class": "PreconditionError"
        },
        "invalid_range_option": {
          "code": 12,
          "message": "Invalid \"range\" option definition.",
          "class": "PreconditionError"
        },
        "invalid_range": {
          "code": 13,
          "message": "Invalid range: min > max.",
          "class": "PreconditionError"
        },
        "invalid_range_type": {
          "code": 14,
          "message": "Invalid \"range.%s\" option: must be of type \"number\".",
          "class": "PreconditionError"
        },
        "invalid_range_format": {
          "code": 15,
          "message": "Option \"%s\": invalid format.",
          "class": "PreconditionError"
        },
        "unable_to_parse_range_value": {
          "code": 16,
          "message": "Unable to parse the range value \"%s\".",
          "class": "PreconditionError"
        },
        "notempty_option_type": {
          "code": 17,
          "message": "Option \"notEmpty\" must be of type \"boolean\".",
          "class": "PreconditionError"
        },
        "missing_values_option": {
          "code": 18,
          "message": "Option \"values\" is required.",
          "class": "PreconditionError"
        },
        "values_option_cannot_be_empty": {
          "code": 19,
          "message": "Option \"values\" must be a non-empty array.",
          "class": "PreconditionError"
        },
        "invalid_values_format": {
          "code": 20,
          "message": "Values must be of type \"string\". Invalid value%s: %s.",
          "class": "PreconditionError"
        },
        "shapetypes_option_cannot_be_empty": {
          "code": 21,
          "message": "Option \"shapeTypes\" must be a non-empty array.",
          "class": "PreconditionError"
        },
        "invalid_shape": {
          "code": 22,
          "message": "Invalid shape%s: %s.",
          "class": "PreconditionError"
        },
        "strict_option_type": {
          "code": 23,
          "message": "Option \"strict\" must be of type \"boolean\".",
          "class": "PreconditionError"
        },
        "invalid_length_option": {
          "code": 24,
          "message": "Invalid \"length\" option definition.",
          "class": "PreconditionError"
        },
        "invalid_length_type": {
          "code": 25,
          "message": "Invalid \"length.%s\" option: must be of type \"number\".",
          "class": "PreconditionError"
        },
        "invalid_length_range": {
          "code": 26,
          "message": "Invalid length range: min > max.",
          "class": "PreconditionError"
        },
        "invalid_properties_for_collection_specification": {
          "code": 27,
          "message": "%s",
          "class": "PreconditionError"
        },
        "structure_collection_validation": {
          "code": 28,
          "message": "%s",
          "class": "BadRequestError"
        },
        "validator_specification_of_collection_error": {
          "code": 29,
          "message": "Validator specification of the collection %s.%s triggered an error",
          "class": "PreconditionError"
        },
        "object_level_undefined_in_specification": {
          "code": 30,
          "message": "All levels of an object have to be defined in the specification.",
          "class": "PreconditionError"
        },
        "throw_error": {
          "code": 31,
          "message": "%s",
          "class": "PreconditionError"
        },
        "manage_error_message": {
          "code": 32,
          "message": "Field %s: %s",
          "class": "BadRequestError"
        },
        "strictness": {
          "code": 33,
          "message": "strictness",
          "class": "BadRequestError"
        },
        "errorcontext": {
          "code": 34,
          "message": "Field %s: %s",
          "class": "BadRequestError"
        }
      }
    },
    "statistics": {
      "code": 7,
      "errors": {
        "invalid_time_value": {
          "code": 1,
          "message": "Invalid time value",
          "class": "BadRequestError"
        }
      }
    },
    "sandbox": {
      "code": 8,
      "errors": {
        "process_already_running": {
          "code": 1,
          "message": "A process is already running for this sandbox",
          "class": "InternalError"
        },
        "timeout": {
          "code": 2,
          "message": "Timeout. The sandbox did not respond within %sms.",
          "class": "GatewayTimeoutError"
        }
      }
    },
    "configuration": {
      "code": 9,
      "errors": {
        "invalid_limits_configuration_format": {
          "code": 1,
          "message": "Invalid config.limits configuration format: please check your Kuzzle configuration files",
          "class": "InternalError"
        },
        "value_out_of_range": {
          "code": 2,
          "message": "Invalid configuration: value set for \"%s\" limit is outside the allowed range",
          "class": "InternalError"
        },
        "concurrentRequests_superior_to_requestsBufferSize": {
          "code": 3,
          "message": "Invalid configuration: the concurrentRequests limit configuration must be strictly inferior to requestsBufferSize",
          "class": "InternalError"
        },
        "requestsBufferWarningThreshold_out_of_range": {
          "code": 4,
          "message": "Invalid configuration: limits.requestsBufferWarningThreshold should be comprised between limits.concurrentRequests and limits.requestsBufferSize",
          "class": "InternalError"
        }
      }
    }
  }
}
